src/v/utils/rwlock.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "seastarx.h"
+#include "ssx/semaphore.h"
+
+#include <seastar/core/rwlock.hh>
+
+namespace ssx {
+
+class rwlock;
+
+class rwlock_unit {
+    rwlock* _lock;
+    bool _is_write_lock{false};
+
+public:
+    rwlock_unit(rwlock* lock, bool is_write_lock) noexcept
+      : _lock(lock)
+      , _is_write_lock(is_write_lock) {}
+    rwlock_unit(rwlock_unit&& token) noexcept
+      : _lock(token._lock)
+      , _is_write_lock(token._is_write_lock) {
+        token._lock = nullptr;
+    }
+    rwlock_unit(const rwlock_unit&) = delete;
+    rwlock_unit(const rwlock_unit&&) = delete;
+    rwlock_unit() = delete;
+
+    ~rwlock_unit() noexcept;
+};
+
+class rwlock : public ss::basic_rwlock<> {
+public:
+    struct lock_result {
+        rwlock_unit unit;
+        bool locked{false};
```

Comment:
I lost your other comment due to force push but std::optional will make the interface cleaner with implicit bool conversion, I'm not sure why seastar was using it, that code probably predates c++-17 that introduced optional.

Reply:
`lock_result` is unused.